src/integTest/kotlin/com/ampnet/payoutservice/blockchain/MerkleTreePathValidatorHashCompatibilityIntegTest.kt

```diff
@@ -0,0 +1,410 @@
+                        it.value.balance.rawValue + BigInteger.ONE,
+                        it.value.treePath(tree)
+                    ).send()
+                ).withIndexedMessage(it.index).isFalse()
+            }
+        }
+    }
+
+    @Test
+    fun mustBeCompatibleWithUnbalancedMerkleTreeWithEvenNumberOfElements() {
+        val accountBalances = listOf(
+            AccountBalance(WalletAddress("0x0"), Balance(BigInteger("0"))),
+            AccountBalance(WalletAddress("0x1"), Balance(BigInteger("1"))),
+            AccountBalance(WalletAddress("0x2"), Balance(BigInteger("2"))),
+            AccountBalance(WalletAddress("0x3"), Balance(BigInteger("3"))),
+            AccountBalance(WalletAddress("0x4"), Balance(BigInteger("4"))),
+            AccountBalance(WalletAddress("0x5"), Balance(BigInteger("5"))),
+            AccountBalance(WalletAddress("0x6"), Balance(BigInteger("6"))),
+            AccountBalance(WalletAddress("0x7"), Balance(BigInteger("7"))),
+            AccountBalance(WalletAddress("0x8"), Balance(BigInteger("8"))),
+            AccountBalance(WalletAddress("0x9"), Balance(BigInteger("9"))),
+            AccountBalance(WalletAddress("0xa"), Balance(BigInteger("10"))),
+            AccountBalance(WalletAddress("0xb"), Balance(BigInteger("11")))
+        )
+        val tree = suppose("Merkle tree with 12 elements is created") {
+            MerkleTree(
+                accountBalances,
+                HashFunction.KECCAK_256
+            )
+        }
+
+        val mainAccount = accounts[0]
+
+        val contract = suppose("Merkle tree path validator is deployed for created tree root") {
+            val future = MerkleTreePathValidator.deploy(
+                hardhatContainer.web3j,
+                mainAccount,
+                DefaultGasProvider(),
+                Numeric.hexStringToByteArray(tree.root.hash.value)
+            ).sendAsync()
+            hardhatContainer.waitAndMine()
+            future.get()
+        }
+
+        verify("contract call will return true for account balances contained in Merkle tree") {
+            accountBalances.withIndex().forEach {
+                assertThat(
+                    contract.containsNode(
+                        it.value.address.rawValue,
+                        it.value.balance.rawValue,
+                        it.value.treePath(tree)
+                    ).send()
+                ).withIndexedMessage(it.index).isTrue()
+            }
+        }
+
+        verify("contract call will return false for account balances not contained in Merkle tree") {
+            accountBalances.withIndex().forEach {
+                assertThat(
+                    contract.containsNode(
+                        it.value.address.rawValue,
+                        it.value.balance.rawValue + BigInteger.ONE,
+                        it.value.treePath(tree)
+                    ).send()
+                ).withIndexedMessage(it.index).isFalse()
+            }
+        }
```

Comment:
Would it make sense to extract this to one function, create and verify, or two functions, create function and verify function?
I'm not a fan of doing this in tests but the same verification process is used in all class tests.

Reply:
Yeah I know... I also don't like doing that in tests but it should probably be fine. If something fails, we will still have test name and index in assert message.